### PR TITLE
Relax concat-linear fusion to support GQA QKV

### PR DESCRIPTION
**Pull request by eble-amd**
## Motivation

The freezing concat-linear optimization fuses independent matmuls that share the same input (e.g., QKV projections) into a single GEMM.  It previously required all weights to have identical shapes, preventing fusion of GQA-style attention where Q has a larger output dimension than K/V (e.g., Q=960, K=V=320).

## Technical Details

Relax the shape check to require only that non-concatenated dimensions match (shape[:-1]), allowing weights and biases to differ in the last dimension.  Use split (which supports unequal sizes) instead of chunk (which requires equal sizes) to split the fused result.

## Test Plan

1. With venv activated,

```shell
LD_LIBRARY_PATH="$(rocm-sdk path --root)/lib:${LD_LIBRARY_PATH:-}" \
    python test/inductor/test_inductor_freezing.py \
    FreezingGpuTests.test_mm_concat_cuda
```

2. Run some system tests with SmolVLA.

## Test Result

- [x] FreezingGpuTests.test_mm_concat_cuda

```
/scratch/eble/pytorch/torch/_inductor/compile_fx.py:320: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
.
----------------------------------------------------------------------
Ran 1 test in 4.402s

OK
```

- [ ] SmolVLA testing

```
TODO
```

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: torch/_inductor/fx_passes/freezing_patterns.py [hunk 1]
```diff
@@ -269,9 +271,10 @@ def addmm_fuse_pattern_second(inp, w1, w2, w3, b1, b2, b3):
         )
 
     def addmm_fuse_replacement_second(inp, w1, w2, w3, b1, b2, b3):
-        cat_w = torch.cat((w1, w2, w3), dim=1)
+        weights = (w1, w2, w3)
+        cat_w = torch.cat(weights, dim=1)
         cat_b = torch.cat((b1, b2, b3))
-        return aten.addmm(cat_b, inp, cat_w).chunk(3, dim=1)
+        return aten.addmm(cat_b, inp, cat_w).split([w.size(1) for w in weights], dim=1)
```
Comment (eble-amd) [inline torch/_inductor/fx_passes/freezing_patterns.py:277]: After writing `shape[:-1]` above, I wondered whether it would be better to write `dim=-1` here and in similar places.

**Comment by eble-amd** [comment]
closing to resubmit upstream

**Comment by eble-amd** [comment]
Upstream PR: https://github.com/pytorch/pytorch/pull/178523
